File: frontend/types/news.ts
// 목적: 금융 뉴스 감성 분석(news) 화면에서 사용하는 타입을 정의한다.
// 주요 역할: Backend GET /news(backend/app/schemas/news.py의 News)의 실제 응답
//           구조를 그대로 미러링한다(2026-08-03 실제 코드 대조 완료).
//
// 주의: Backend는 뉴스 한 건에 별도 id를 내려주지 않는다(News Response Schema에
// 의도적으로 id가 빠져 있음 — backend/app/schemas/news.py 주석 참고). 화면에서
// 뉴스 한 건을 식별할 때는 원문 URL(url)을 고유 키로 쓴다.

export type NewsBriefType = "morning" | "closing";

// backend/app/clients/gemma_client.py의 SENTIMENT_SCHEMA가 정한 값 그대로다.
// 요약·판정 배치가 아직 처리하지 않은 기사는 null로 내려온다.
export type NewsLabel = "POS" | "NEU" | "NEG";

export interface NewsEvidence {
  sentence: string;
  is_quote: boolean;
}

export interface NewsArticle {
  market_date: string;
  brief_type: NewsBriefType;
  title: string;
  published_at: string;
  url: string;
  source: string;
  body: string;
  summary: string | null;
  label: NewsLabel | null;
  evidence: NewsEvidence[];
}

File: backend/app/services/news.py
"""news.py
레이어: Services
역할: 뉴스 원문과 한 줄 요약, 감성 판정 근거를 하나의 조회 결과로 합친다.
"""

from collections import defaultdict

from ..repositories import news_article as news_article_repository
from ..schemas.news import KST, News, NewsEvidence


def build_news_list(articles: list[dict], evidences: list[dict]) -> list[News]:
    """기사 목록에 근거 문장을 기사별로 묶어 붙인다."""

    grouped: dict[int, list[NewsEvidence]] = defaultdict(list)
    for evidence in evidences:
        grouped[evidence["article_id"]].append(
            NewsEvidence(
                sentence=evidence["sentence"],
                is_quote=evidence["is_quote"],
            )
        )

    return [
        News(
            market_date=article["market_date"],
            brief_type=article["brief_type"],
            title=article["title"],
            # market_date 가 KST 기준이므로 게시시각도 같은 기준으로 맞춘다.
            # Driver 가 돌려주는 UTC 를 그대로 내보내면 날짜가 하루 어긋날 수 있다.
            published_at=article["published_at"].astimezone(KST),
            url=article["url"],
            source=article["source"],
            body=article["body"],
            summary=article["summary"],
            label=article["label"],
            evidence=grouped.get(article["id"], []),
        )
        for article in articles
    ]


def read_news() -> list[News]:
    """전체 뉴스를 요약·판정과 함께 최신순으로 조회한다."""

    articles = news_article_repository.find_news()
    if not articles:
        return []

    evidences = news_article_repository.find_verified_evidences()
    return build_news_list(articles, evidences)
